### PR TITLE
cli: add command to compile libraries

### PR DESCRIPTION
**Pull request by hackaugusto**
## Describe your changes

This adds a CLI command to compile masm files to a masl, useful for CLI users.

## Checklist before requesting a review
- Repo forked and branch created from `next` according to naming convention.
- Commit messages and codestyle follow [conventions](./CONTRIBUTING.md).
- Relevant issues are linked in the PR description.
- Tests added for new functionality.
- Documentation/comments updated according to changes.

### CODE REVIEW

File: miden/src/cli/compile_lib.rs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+#[derive(StructOpt, Debug)]
+#[structopt(name = "Compile Library", about = "Compiles a library")]
+pub struct CompileLibraryCmd {
+    /// Path to a directory containg the `.masm` files which are part of the library
+    #[structopt(parse(from_os_str))]
+    dir: PathBuf,
+    /// Name of the top-level namespace, e.g. `mylib`.
+    #[structopt()]
+    namespace: String,
+}
```
Comment (hackaugusto) [inline miden/src/cli/compile_lib.rs:14]: I tried to reuse the `CompileCmd`, but `compile` is written to `compile` a whole program, which may use multiple libraries, each with their own namespace, i couldn't think of a simple/clear interface to do that.

**Review by bobbinth — CHANGES_REQUESTED** [review]
Looks good! Thank you! I left a few comments inline. The main thing is that I'm thinking we should change the command to work more like this:
```
miden bundle <dir> --version <version> --namespace <namespace>
```
Where `namespace` is optional.

File: miden/src/cli/compile_lib.rs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+impl CompileLibraryCmd {
+    pub fn execute(&self) -> Result<(), String> {
+        println!("============================================================");
+        println!("Compile library");
+        println!("============================================================");
+
+        let namespace =
+            LibraryNamespace::try_from(self.namespace.to_string()).expect("invalid base namespace");
+        let version = Version::try_from(env!("CARGO_PKG_VERSION")).expect("invalid cargo version");
```
[the remark this answers is not in the document]
Comment (hackaugusto) [reply miden/src/cli/compile_lib.rs:24]: added a CLI optional argument, set the default to `0.1`, let me know if it should instead be a required positional argument.

File: miden/src/cli/compile_lib.rs [hunk 3]
```diff
@@ -0,0 +1,36 @@
+use assembly::{LibraryNamespace, MaslLibrary, Version};
+use std::path::PathBuf;
+use structopt::StructOpt;
+
+#[derive(StructOpt, Debug)]
+#[structopt(name = "Compile Library", about = "Compiles a library")]
+pub struct CompileLibraryCmd {
```
[the remark this answers is not in the document]
Comment (hackaugusto) [reply miden/src/cli/compile_lib.rs:7]: changed.

**Comment by hackaugusto** [comment]
> Looks good! Thank you! I left a few comments inline. The main thing is that I'm thinking we should change the command to work more like this:

This would force the user to have a different folder per bundle, that is a more constrained version of the command. For example, if there are two versions of the same library, and they share a `utils`, that design would require copying the code.

Should I change?

**Comment by bobbinth** [comment]
> This would force the user to have a different folder per bundle, that is a more constrained version of the command. For example, if there are two versions of the same library, and they share a `utils`, that design would require copying the code.

Not sure I fully followed this: doesn't the command take all MASM files from a given directory? How would a shared directory be specified in this case? (and what namespace in the library hierarchy would it get?)

**Comment by hackaugusto** [comment]
> Not sure I fully followed this: doesn't the command take all MASM files from a given directory? How would a shared directory be specified in this case? (and what namespace in the library hierarchy would it get?)

My mistake. I had something else in mind. I'm making the requested changes.

**Review by bobbinth — APPROVED** [review]
Looks good! Thank you! I added one commit which fixes a few typos in the docs.

**Pull request merged**